lib/Serializer.ts

```diff
@@ -0,0 +1,119 @@
+import { ObjectType, ModelType } from "./General";
+import { Document } from "./Document";
+import CustomError = require("./Error");
+import utils = require("./utils");
+
+export interface SerializerOptions {
+	include?: string[];
+	exclude?: string[];
+	modify?: (serialized: ObjectType, original: ObjectType) => ObjectType;
+}
+
+export class Serializer {
+	#serializers: {[key: string]: SerializerOptions};
+	#defaultSerializer: string;
```

Comment:
I'm just curious, why not TypeScript's `protected`/`private` for this? It accomplishes the same goal for the API, but leaves some leeway for people who _really_ want to access these. You never know all people's requirements beforehand, and I don't see any harm in enforcing this on the type-checker level instead.

Reply:
@dolsem I think this falls under undocumented things shouldn't be used. I'd love to do this more throughout the code to try to enforce it more. Once plugin support is added, some of those things will probably be reversed and will be more open. But I'd like to get to the point where internal things are **impossible** to access and are marked somehow in the codebase. Anything that is modified that isn't internal or private, should be considered a breaking change. At that point we'd be able to remove that documentation note from the website. But modifying internal things would be allowed and wouldn't be considered a breaking change.

https://dynamoosejs.com/other/FAQ#can-i-use-an-undocumented-property-class-method-or-function-in-dynamoose

Reply:
It would also be great at some point to ensure that tests conform to that. So anytime you have to modify or delete a test, that is considered a breaking change. However adding tests is completely allowed.

We aren't quite at that point yet, and there are some cases where the tests are just wrong. So bug fixes require changes to the tests.

So the goal and idea of this is to work towards enforcing more backwards compatibility and ensuring there are less breaking changes (or at least that they are marked and detected correctly).

Reply:
But why prevent people from accessing Dynamoose's internals if they really want to? They might have good reasons for it, and it will be their responsibility to deal with breaking changes since these things are technically not supposed to be accessed. Just like there's reflection in Java, I think there should be a way in JS/TS to access internals if the user can't work around that.

Reply:
@dolsem I assume when using 3rd party packages, that when printing out something, that I'm able to access and use everything it displays. In this case if you do `console.log(model.serializer)` it prints out basically an empty object. However, if we were to just set these to TS `private` for JS users there is no signal that these are private properties.

We could add `_` as a prefix. However I'm not so sure that is the best indication anymore.

See my first bullet point here: https://github.com/dynamoose/dynamoose/issues/900#issuecomment-627652795.

So I think there are users who are using internal properties in ways they shouldn't be.

Now if it gets to the point where users wish to access this, or feel like there is something they want to do that they can't. Then we'll cross that bridge when we get to it. But I think the pros of preventing access outweigh the drawbacks. Especially in this case I can't think of any conceivable reason why you'd need to access these things. Maybe there is, but if there is we should create a standard API and have a better way to do it that is documented.

In that comment above, it was a clear indication of the user wanting to do something that wasn't yet supported. Passing a schema into another schema. Which is something I'm working on implementing. But normally when users reach for internal properties, it's an indication that Dynamoose could do something better or isn't doing something correctly. Which should be fixed as opposed to just allowing access to internal things IMO.

Reply:
@fishcharlie yeah, you're right. Forgot about JS for a second.